clippy_lints/src/methods/unnecessary_fallible_conversions.rs

```diff
@@ -0,0 +1,106 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::get_parent_expr;
+use clippy_utils::ty::implements_trait;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+use rustc_span::{sym, Span};
+
+use super::UNNECESSARY_FALLIBLE_CONVERSIONS;
+
+/// What function is being called and whether that call is written as a method call or a function
+/// call
+#[derive(Copy, Clone)]
+#[expect(clippy::enum_variant_names)]
+enum FunctionKind {
+    /// `T::try_from(U)`
+    TryFromFunction,
+    /// `t.try_into()`
+    TryIntoMethod,
+    /// `U::try_into()`
+    TryIntoFunction,
+}
+
+fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &Expr<'_>,
+    node_args: ty::GenericArgsRef<'tcx>,
+    kind: FunctionKind,
+    primary_span: Span,
+) {
+    if let &[self_ty, other_ty] = node_args.as_slice()
+        // useless_conversion already warns `T::try_from(T)`, so ignore it here
+        && self_ty != other_ty
+        && let Some(self_ty) = self_ty.as_type()
+        && let Some(from_into_trait) = cx.tcx.get_diagnostic_item(match kind {
+            FunctionKind::TryFromFunction => sym::From,
+            FunctionKind::TryIntoMethod | FunctionKind::TryIntoFunction => sym::Into,
+        })
+        // If `T: TryFrom<U>` and `T: From<U>` both exist, then that means that the `TryFrom`
+        // _must_ be from the blanket impl and cannot have been manually implemented
+        // (else there would be conflicting impls, even with #![feature(spec)]), so we don't even need to check
+        // what `<T as TryFrom<U>>::Error` is.
+        && implements_trait(cx, self_ty, from_into_trait, &[other_ty])
+        // Avoid linting if this is an argument to a function: `f(i32.try_into())`.
+        // The function expects a `Result` either way, so users would need to wrap it back in `Ok(..)`
+        // and save on nothing
+        && get_parent_expr(cx, expr).map_or(true, |parent| !matches!(parent.kind, ExprKind::Call(..)))
+    {
+        span_lint_and_sugg(
+            cx,
+            UNNECESSARY_FALLIBLE_CONVERSIONS,
+            primary_span,
+            match kind {
+                FunctionKind::TryFromFunction => "calling `TryFrom::try_from` when `From` could be used",
+                FunctionKind::TryIntoMethod
+                | FunctionKind::TryIntoFunction => "calling `TryInto::try_into` when `Into` could be used",
+            },
+            "use",
+            match kind {
+                FunctionKind::TryFromFunction => "From::from".into(),
+                FunctionKind::TryIntoFunction => "Into::into".into(),
+                FunctionKind::TryIntoMethod => "into".into(),
+            },
+            // Suggestion likely results in compile errors, so it needs to be adjusted a bit by the user
+            // (e.g. removing `unwrap()`s as a result, changing type annotations, ...)
```

Comment:
Maybe the lint could detect specifically `.try_into().unwrap()` and `.try_into().expect()` and lint those with `Applicability::MachineApplicable`.